Mathlib/Algebra/Ring/Associator.lean

```diff
@@ -0,0 +1,107 @@
+/-
+Copyright (c) 2025 Bernhard Reinke. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bernhard Reinke
+-/
+import Mathlib.Algebra.Ring.Defs
+import Mathlib.Algebra.Group.Hom.End
+import Mathlib.Algebra.Group.Hom.Instances
+import Mathlib.Tactic.Abel
+
+/-!
+## Associator in a ring
+
+If `R` is a non-associative ring, then  `(x * y) * z - x * (y * z)` is called the `associator` of
+ring elements `x y z : R`.
+
+The associator vanishes exactly when `R` is associative.
+
+We prove variants of this statement also for the `AddMonoidHom` bundled version of the associator,
+as well as the bundled version of `mulLeft₃` and `mulRight₃`, the multiplications `(x * y) * z` and
+`x * (y * z)`.
+-/
+
+universe u
+
+section NonUnitalNonAssocRing
+variable {R : Type u} [NonUnitalNonAssocRing R]
+
+/-- The associator `(x * y) * z - x * (y * z)` -/
+def associator (x y z : R) := (x * y) * z - x * (y * z)
+
+theorem associator_eq_mulLeft₃_sub_mulRight₃ : associator = mulLeft₃ (G := R) - mulRight₃ := by
+  funext x y z
+  simp [associator]
+
+theorem associator_eq_zero_iff_associative :
+    associator (R := R) = 0 ↔ Std.Associative (fun (x y : R) ↦ x * y) := by
+  simp [mulLeft₃_eq_mulRight₃_iff_associative, associator_eq_mulLeft₃_sub_mulRight₃, sub_eq_zero]
+
+theorem associator_cocycle (a b c d : R) :
+    a * associator b c d - associator (a * b) c d + associator a (b * c) d - associator a b (c * d)
+    + (associator a b c) * d = 0 := by
+  simp only [associator, mul_sub, sub_mul]
+  abel1
+
+end NonUnitalNonAssocRing
+
+namespace AddMonoidHom
+
+section NonUnitalNonAssocSemiring
+variable {R : Type u} [NonUnitalNonAssocSemiring R]
+
+/-- The multiplication `(x * y) * z` of three elements of a (non-associative)
+(semi)-ring is an `AddMonoidHom` in each argument. -/
+def mulLeft₃ : R →+ R →+ R →+ R where
+  toFun x := comp mul (mulLeft x)
+  map_zero' := by ext; simp
+  map_add' x y := by ext; simp [add_mul]
+
+@[simp]
+theorem mulLeft₃_apply (x y z : R) : mulLeft₃ x y z = (x * y) * z := rfl
+
+/-- The multiplication `x * (y * z)` of three elements of a (non-associative)
+(semi)-ring is an `AddMonoidHom` in each argument. -/
+def mulRight₃ : R →+ R →+ R →+ R where
+  toFun x := compr₂ mul (mulLeft x)
+  map_zero' := by ext; simp
+  map_add' x y := by ext; simp [add_mul]
```

Comment:
One of these is related to [`LinearMap.mulLeftRight`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Bilinear.html#LinearMap.mulLeftRight); could you add a "see also" to the docstring?